### PR TITLE
Implement backend pagination for GET /listings

### PR DESCRIPTION
**Pull request by anders-schneider**
## Issue

Addresses #56 
- [X] This change addresses the issue in full

## Description

This change changes the parameters and return type for `GET /listings` such that results are paginated. The parameters now include `page` and `limit` fields, and the returned object includes both an array of Listings and metadata about e.g. which page of listings this represents, how many total listings there are, how many listings per page, etc.

The pagination implementation uses `skip` and `take` when building the SQL query, and those correspond to SQL's `OFFSET` and `LIMIT`, respectively. `typeorm` actually sends two queries as a result: a first to ask for all the IDs of listings that are in the specified range, and a second to get all of the data (including all of the data in `JOIN`ed tables). (And, this change adds a third `SELECT COUNT(*) FROM listings;` query in order to populate the pagination metadata, but that should be fast.) The hope is that, although this change means more SQL queries to the database, the overall time to retrieve listing data is faster since we don't have to retrieve all of the `JOIN`ed data in a single query.

This change also adds a couple tests and fixes a couple places in the frontend so that they understand the new paginated-listings return type.

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How Can This Be Tested/Reviewed?

```
$ cd backend/core
$ yarn test:e2e:local
$ yarn test listing
```

Also, if you run the code locally and point your browser to `localhost:3100/listings`, you should be able to see the new paginated listings result.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [ ] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have updated the changelog to include a description of my changes

### CODE REVIEW

File: backend/core/src/listings/listings.service.ts [hunk 1]
```diff
@@ -36,12 +40,61 @@ export class ListingsService {
...
+
+    // Sort units and preferences.
+    // This step was removed from the SQL query because it interferes with pagination
+    // (See https://github.com/CityOfDetroit/affordable-housing-app/issues/88#issuecomment-865329223)
+    listings.forEach((listing) => {
+      listing.property.units.sort((a, b) => a.maxOccupancy - b.maxOccupancy)
+      listing.preferences.sort((a, b) => a.ordinal - b.ordinal)
+    })
+
+    // TODO: remove jsonpath filtering (it doesn't play well with pagination)
```
Comment (willrlin) [inline backend/core/src/listings/listings.service.ts:81]: Can you explain more about this?

File: backend/core/src/listings/listings.controller.ts [hunk 2]
```diff
@@ -60,14 +65,9 @@ export class ListingsController {
...
-  @ApiImplicitQuery({
-    name: "jsonpath",
-    required: false,
-    type: String,
-  })
   @UseInterceptors(CacheInterceptor)
-  async list(@Query() queryParams: ListingsListQueryParams): Promise<ListingDto[]> {
-    return mapTo(ListingDto, await this.listingsService.list(queryParams))
+  async list(@Query() queryParams: ListingsListQueryParams): Promise<PaginatedListingsDto> {
+    return await this.listingsService.list(queryParams)
```
Comment (willrlin) [inline backend/core/src/listings/listings.controller.ts:70]: Do you ever use the PaginatedListings object? Could you just unwrap it here and not have to change the API so it still returns Promise<ListingDto[]>?

File: backend/core/src/listings/listings.controller.ts [hunk 3]
```diff
@@ -60,14 +65,9 @@ export class ListingsController {
 
   @Get()
   @ApiOperation({ summary: "List listings", operationId: "list" })
-  @ApiImplicitQuery({
-    name: "jsonpath",
-    required: false,
-    type: String,
-  })
```
Comment (willrlin) [inline backend/core/src/listings/listings.controller.ts:67]: So you're removing this capability entirely?

File: backend/core/src/listings/listings.service.ts [hunk 4]
```diff
@@ -36,12 +40,61 @@ export class ListingsService {
       query.andWhere("property.neighborhood = :neighborhood", { neighborhood: params.neighborhood })
     }
 
-    let listings = await query.getMany()
+    let currentPage: number = params.page
+    let itemsPerPage: number = params.limit
```
Comment (willrlin) [inline backend/core/src/listings/listings.service.ts:44]: What happens if there's no page or limit param? Do they default to a value or does this just fail?

**Review by willrlin — CHANGES_REQUESTED** [review]

**Comment by willrlin** [comment]
Also, change the title of the PR to imperative and remove your username, as if it were the first line of the CL.

File: backend/core/src/listings/listings.service.ts (hunk 1 above)
Comment (anders-schneider) [reply backend/core/src/listings/listings.service.ts]: Yeah sorry, the comment was too short - I updated it to say that if the jsonpath filtering removes any listings, the pagination metadata that's returned will be incorrect.

File: backend/core/src/listings/listings.controller.ts (hunk 2 above)
Comment (anders-schneider) [reply backend/core/src/listings/listings.controller.ts:70]: We don't yet use the PaginatedListingsDto, but we will want to in the /listings frontend: we'll want to display what page we're on, whether there is a next page (or we're on the last page), maybe how many total listings there are across all pages.

File: backend/core/src/listings/listings.controller.ts (hunk 3 above)
Comment (anders-schneider) [reply backend/core/src/listings/listings.controller.ts:67]: I think this should have been removed/changed in https://github.com/CityOfDetroit/bloom/pull/18. I looked into whether Bloom adds this ApiImplicitQuery (or the new ApiQuery) decorator anywhere else, and it doesn't look like it.

File: backend/core/src/listings/listings.service.ts (hunk 4 above)
Comment (anders-schneider) [reply backend/core/src/listings/listings.service.ts:44]: The default values are `page: 1` and `limit: 10` (https://github.com/CityOfDetroit/bloom/blob/ec9124b6281953e2c5b1ac63a83d38e5ceb3cad5/backend/core/src/shared/dto/pagination.dto.ts#L44). It's possible to set them to something else in the request, but if they are invalid values (e.g. if either is <= 0) we just return all listings.

File: backend/core/src/listings/listings.controller.ts (hunk 2 above)
Comment (abbiefarr) [reply backend/core/src/listings/listings.controller.ts:69]: Will this also force us to change any implementation in the partner portal that fetches listings? E.g will we have to incorporate pagination into the partners' page to view their own listings?

**Comment by anders-schneider** [comment]
Thanks for the comments @willrlin ! I closed the conversations that I felt were resolved, and left the ones that you may have follow-up on open - let me know? And I changed the PR title, thanks for catching that!

File: backend/core/src/listings/listings.controller.ts (hunk 2 above)
Comment (anders-schneider) [reply backend/core/src/listings/listings.controller.ts:69]: Yeah good call, we will have to update the parts of the partner portal that call `GET /listings` (similar to the one-line changes I made to the `/sites/public/pages` files in this PR). But if we don't want to do the frontend work to the partner portal to add pagination, we can just explicitly set the params on the `GET /listings` requests to e.g. `page: -1, limit: -1` and that will make it so that the results include all listings.

TLDR: we'll need to make a couple one-line changes to merge Exygy's partner-portal-listings work into our repo.

Comment (abbiefarr) [reply backend/core/src/listings/listings.controller.ts:69]: This is probably a big ask, but it may be worth it to decouple the queries so we have the ability to request both paginated and unpaginated versions of listings. WDYT?

Comment (willrlin) [reply backend/core/src/listings/listings.controller.ts:69]: Or you could make the defaults `page: -1, limit: -1` such that `GET /listings` will return the same as it would today?

File: backend/core/src/listings/listings.controller.ts (hunk 3 above)
Comment (willrlin) [reply backend/core/src/listings/listings.controller.ts:67]: I see, so this wasn't Exygy that added it, but us?

File: backend/core/src/listings/listings.service.ts (hunk 1 above)
Comment (willrlin) [reply backend/core/src/listings/listings.service.ts]: Just so I am 100% understanding this - is this jsonpath filtering something that Exygy has built in Bloom? Could we later incorporate that into the DB query so it still works instead of post-DB query? Or does the pagination library fail if you try to limit the query?

File: backend/core/src/listings/listings.controller.ts (hunk 2 above)
Comment (abbiefarr) [reply backend/core/src/listings/listings.controller.ts:69]: @willrlin that seems reasonable, given that AFAICT pagination is currently only used for applications, which we're not using anyway.

Comment (anders-schneider) [reply backend/core/src/listings/listings.controller.ts:69]: Thanks for these suggestions! Just pushed a commit that makes the default values `page: -1, limit: -1`. I had to make changes to the ApplicationService so that the negative values wouldn't break it. With those changes, both `yarn test` and `yarn test:e2e:local` pass.

File: backend/core/src/listings/listings.controller.ts (hunk 3 above)
Comment (anders-schneider) [reply backend/core/src/listings/listings.controller.ts:67]: This `ApiImplicitQuery` decorator was there in the beginning (when `jsonpath` was the only variable in this method), when we forked Exygy's repo. Now that #18 made `jsonpath` just a field within `ListingsListQueryParams`, this decorator is incorrect

File: backend/core/src/listings/listings.service.ts (hunk 1 above)
Comment (anders-schneider) [reply backend/core/src/listings/listings.service.ts]: We chatted about this offline, recording a couple notes:
- Exygy built in the jsonpath filtering support, but they don't seem to be using it anywhere outside of tests
- We don't see a use for `jsonpath` since our SQL-level filtering will give us all the filtering capabilities we want
- This change leaves `jsonpath` filtering in place, though, so that we can think more about if we want to remove it later, or maybe keep it for the sake of keeping this part of our codebase similar to Exygy's

Comment (anders-schneider) [reply backend/core/src/listings/listings.service.ts]: I filed https://github.com/CityOfDetroit/bloom/issues/135 and pointed the TODO at that

**Review by abbiefarr — CHANGES_REQUESTED** [review]
This change seems to break both the partner portal:
image

and the resident portal:
image

For the partner portal, I'm hoping it's an easy fix in sites/partners/lib/hooks.ts. I have to look more into the issue with the resident portal.

**Comment by abbiefarr** [comment]
Also I believe that whenever the backend api changes, we need to regenerate the client with `yarn generate:client`

**Comment by abbiefarr** [comment]
I fixed the errors with the public and partner portals, with the default values they display all the listings as they did before. I also ran the client library codegen, which generated some changes I expected and others I didn't expect and can't really explain.

PTAL at the newest commit, the issues were with some objects being "undefined", but I'm pretty unfamiliar with typescript so I'm pretty skeptical of the code I wrote.

File: backend/core/src/listings/listings.controller.ts (hunk 2 above)
Comment (willrlin) [reply backend/core/src/listings/listings.controller.ts:69]: Hm, so does that mean that the partner portal frontend is not calling with page/limit because they're using the default? Maybe we should file a task to consider putting 10 as the default limit in `ApplicationService` (i.e. `limit > 0 ? limit : 10`) instead of max integer.

File: sites/public/pages/index.tsx [hunk 5]
```diff
@@ -82,7 +82,7 @@ export async function getStaticProps() {
   let listings = []
   try {
     const response = await axios.get(process.env.listingServiceUrl)
-    listings = response.data
+    listings = response.data.items ? response.data.items : null
```
Comment (willrlin) [inline sites/public/pages/index.tsx:85]: You have `data.items ?` here and `data ? data.items : []` above. Could there exist a case where data is valid but data.items is not?

**Review by willrlin — APPROVED** [review]
Seems fine to me. As for the codegen, hopefully it was just a bit out of date. Maybe we should set a presubmit check to ensure that the generated version matches the PR version?

File: sites/public/pages/index.tsx (hunk 5 above)
Comment (abbiefarr) [reply sites/public/pages/index.tsx:85]: I don't think so, it looks like `items` is a required property of the `Pagination` object. Will update to check `response.data` here.

File: backend/core/src/listings/listings.controller.ts (hunk 2 above)
Comment (abbiefarr) [reply backend/core/src/listings/listings.controller.ts:69]: Updated to make the default in ApplicationService 10 instead of max integer.

**Review by abbiefarr — APPROVED** [review]
